all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+  /// when there are no idle workflows in the cache (IE: All other outstanding slots are in use), we will buffer the
+  /// task and wait for another to complete so we can evict it and make room for the new one.
+  fn can_allow_workflow(&self, slots_info: &WorkflowSlotsInfo, new_task: &WorkflowSlotInfo) -> bool;
+  /// Called when deciding what workflow should be evicted from the cache. May return the run id of a workflow to evict.
+  /// If None, or an invalid run id is returned, the default LRU eviction / strategy is used. The id returned must
+  /// be one of the workflows provided in `evictable` (the ones for which there is no actively processing WFT).
+  fn eviction_hint(&self, evictable: &[WorkflowSlotInfo]) -> Option<&str>;
+  /// Called when a workflow is evicted from the cache
+  fn evicted_workflow(&self, evicted: &WorkflowSlotInfo);
+  /// If there is a fixed maximum number of cached workflows, return it.
+  fn max(&self) -> Option<usize>;
+}
+
+/// The user would be allowed to provide either just a slot supplier (using the standard fixed-size cache),
+/// or one of these to handle both themselves.
+/// 
+/// I'm a bit at a loss for a good name
+trait WorkflowSlotAndCacheManager: WorkflowTaskSlotSupplier + WorkflowCacheSizer {}
+```
+
+## Flow
+
+Some context on the current flow (at least in Core). The cache only applies to workflow tasks. This flow should remain
+the same after these changes. The flow makes it clear that a misbehaving `WorkflowSlotAndCacheManager` which allows lots
+of slots but keeps saying the cache should not expand could result in buffering tasks and generally gumming things
+up.
+
+```mermaid
+sequenceDiagram
+  participant T as Temporal
+  participant W as Worker
+  participant S as Slots
+  participant C as CacheSizer
+  W->>S: reserve_slot
+  S-->>W: Allowed
+  W->>T: Poll
+  T-->>W: Response
+  
+  alt Workflow is not already in cache
+    W->>C: `can_allow_workflow` ?
+    C-->>W: I'm full
+    W->>W: Buffer task
+    W->>W: Wait for some workflow to be idle (maybe there is already)
+    W->>C: `eviction_hint`
+    C-->>W: Evict This guy
+    W->>W: Evict workflow
+    W->>C: `evicted_workflow`
+    W->>W: Cache new workflow
+  end
+  
+  W->>S: `mark_slot_used`
+  W->>W: Process workflow task
+  W->>S: `release_slot`
+```
+
+## Default implementations
+
+We should provide a few default implementations:
+
+* Semaphore based SlotSupplier: This is the current implementation. There's a fixed number of slots, that's it.
+* Resource based SlotSupplier: This implementation can look at how close the worker is to some resource limit
+  (% of CPU and/or memory used), and hand out slots if it's still below that limit. This only makes sense for
+  activity slot suppliers, since workflows use little CPU, and the memory they use is the concern of cache
+  management, not the slot supplier.
+* Memory based WorkflowCacheSizer: Like above, but for the sizing of the workflow cache.
```

Comment:
We should get into the details somewhere about these impls, e.g. whether it's process mem used or system-free-mem based, encouragement of cross-worker use, which CPU load number is referenced and how it's aggregated across cores, what config values are exposed, etc.

I am ok deferring post-proposal, but I would like SDKs to be consistent in behavior and config.

Reply:
Yeah, I agree. I just didn't want to have this get cluttered with the details of that.